pkg/expapi/types.go

```diff
@@ -134,3 +136,40 @@ type HorizontalPodAutoscalerList struct {
 
 	Items []HorizontalPodAutoscaler `json:"items"`
 }
+
+type DerivedNodeMetrics struct {
+	api.TypeMeta   `json:",inline"`
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	NodeMetrics      MetricsWindows            `json:"nodemetrics"`
+	SystemContainers []DerivedContainerMetrics `json:"systemcontainers"`
+}
+
+type DerivedContainerMetrics struct {
+	Name    string         `json:"name"`
+	Windows MetricsWindows `json:"windows"`
+}
+
+type DerivedNodeMetricsList struct {
+	api.TypeMeta `json:",inline"`
+	api.ListMeta `json:"metadata,omitempty"`
+
+	Items []DerivedNodeMetrics `json:"items"`
+}
+
+// Last overlapping 10s, 1m, 1h and 1d as a start
+// Updated every 10s, so the 10s window is sequential and the rest are
+// rolling.
+type MetricsWindows map[time.Duration]DerivedMetrics
+
+type DerivedMetrics struct {
+	// End time of all the time windows in Metrics
+	EndTime util.Time `json:"endtime"`
```

Comment:
Generally we distinguish between *util.Time and util.Time for whether a
value is present or not.  If EndTime *must* be set, then what you have is
correct.  But is EndTime *required* by validation?

On Fri, Aug 21, 2015 at 4:49 PM, Daniel Martí <notifications@github.com>
wrote:

> In pkg/expapi/types.go
> <https://github.com/kubernetes/kubernetes/pull/12952#discussion_r37676783>
> :
>
> > +
> > +type DerivedNodeMetricsList struct {
> > +	api.TypeMeta `json:",inline"`
> > +	api.ListMeta `json:"metadata,omitempty"`
> > +
> > +	Items []DerivedNodeMetrics `json:"items"`
> > +}
> > +
> > +// Last overlapping 10s, 1m, 1h and 1d as a start
> > +// Updated every 10s, so the 10s window is sequential and the rest are
> > +// rolling.
> > +type MetricsWindows map[time.Duration]DerivedMetrics
> > +
> > +type DerivedMetrics struct {
> > +	// End time of all the time windows in Metrics
> > +	EndTime util.Time `json:"endtime"`
>
> If it could be empty, it would be json:"endtime,omitempty".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12952/files#r37676783>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift